File: Difficulty: Medium/Replace O's with X's/replace-os-with-xs.cpp
// User function Template for C++


#define graph_t vector<vector<char>>
#define visit_t vector<vector<bool>>

class Solution {
  public:
    vector<int> dirx = {0,0,1,-1};
    vector<int> diry = {1,-1,0,0};
    bool valid(int i ,int j, visit_t& visited, graph_t& mat) {
        int n = mat.size();
        int m = mat[0].size();
        if(i>=0 && i<n && j>=0 && j<m && mat[i][j]=='O' && visited[i][j] == false)
            return true;
        return false;
    }
    vector<vector<char>> fill(vector<vector<char>>& mat) {
        // code here
        int n = mat.size();
        int m = mat[0].size();
        
        visit_t visited(n,vector<bool>(m,false));
        queue<pair<int,int>> que;
        
        for(int i = 0;i<n;i++) {
            if(mat[i][0]=='O') {
                que.push({i,0});
                visited[i][0] = true;
            }
            if(mat[i][m-1]=='O') {
                que.push({i,n-1});
                visited[i][n-1]= true;
            }
        }
        
        for(int i = 1;i<m-1;i++) {
            if(mat[0][i]=='O') {
                que.push({0,i});
                visited[0][i] = true;
            }
            if(mat[n-1][i]=='O') {
                que.push({n-1,i});
                visited[n-1][i] = true;
            }
        }
        
        while(!que.empty()) {
            pair<int,int> nd = que.front();
            que.pop();
            
            for(int i = 0;i<4;i++) {
                int newi = nd.first+dirx[i];
                int newj = nd.second+diry[i];
                
                if(valid(newi,newj,visited,mat)) {
                    visited[newi][newj] = true;
                    que.push({newi,newj});
                }
            }
            
        }
        
        for(int i = 0 ;i<n;i++) {
            for(int j = 0;j<m;j++) {
                if(visited[i][j]==false && mat[i][j]=='O') {
                    mat[i][j] = 'X';
                }
            }
        }
        
        
        return mat;
    }
};

File: Difficulty: Medium/Replace O's with X's/replace-os-with-xs.java
class Solution {
    public void fill(char[][] grid) {
        // Code here
        int n = grid.length;
        int m = grid[0].length;
        Deque<int[]> que = new ArrayDeque<int[]>();
        
        for(int i = 0;i<n;i++) {
            if(i==0 || i==n-1) {
                for(int j = 0;j<m;j++) {
                    if(grid[i][j]=='O') {
                        que.add(new int[]{i,j});
                        grid[i][j] = 'B';
                    }
                }
            }
            else if(grid[i][0]=='O' || grid[i][m-1]=='O') {
                if(grid[i][0]=='O') {
                    que.add(new int[]{i,0});
                    grid[i][0] = 'B';
                }
                if(grid[i][m-1]=='O') {
                    que.add(new int[]{i,m-1});
                    grid[i][m-1] = 'B';
                }
            }
        }
        
        int[] dx = {0,1,0,-1};
        int[] dy = {1,0,-1,0};
        
        while(que.isEmpty()!=true) {
            int[] node = que.poll();
            for(int i = 0;i<4;i++) {
                int x = node[0]+dx[i];
                int y = node[1]+dy[i];
                if(x>=0 && x<n && y>=0 && y<m && grid[x][y]=='O') {
                    grid[x][y] = 'B';
                    que.add(new int[]{x,y});
                }
            }
        }
        
        for(int i = 0;i<n;i++) {
            for(int j = 0;j<m;j++) {
                if(grid[i][j]=='O') {
                    grid[i][j] = 'X';
                }
            }
        }
        
        for(int i = 0;i<n;i++) {
            for(int j = 0;j<m;j++) {
                if(grid[i][j]=='B') {
                    grid[i][j] = 'O';
                }
            }
        }
        
    }
}
